implementations/rust/ockam/ockam_identity/src/secure_channel/encryptor_worker.rs

```diff
@@ -91,33 +141,154 @@ impl EncryptorWorker {
+                    self.schedule_credentials_refresh(ctx, true).await?;
+                    return Err(err);
+                }
             }
+        } else {
+            return Err(IdentityError::NoCredentialsRetriever.into());
+        };
+
+        let versioned_data = credential.credential.get_versioned_data()?;
+        let data = CredentialData::get_data(&versioned_data)?;
+        self.min_credential_expiration = Some(data.expires_at);
+
+        let msg = RefreshCredentialsMessage {
+            change_history,
+            credentials: vec![credential],
         };
+        let msg = SecureChannelMessage::RefreshCredentials(msg);
+
+        let msg = self.encrypt(ctx, msg).await?;
+
+        info!(
+            "Sending credentials refresh for {}",
+            self.addresses.encryptor
+        );
 
         // Send the message to the decryptor on the other side
         ctx.send_from_address(
             self.remote_route.clone(),
-            encrypted_payload,
+            msg,
             self.addresses.encryptor.clone(),
         )
         .await?;
 
+        self.schedule_credentials_refresh(ctx, false).await?;
+
+        Ok(())
+    }
+
+    async fn send_close_channel(&mut self, ctx: &Context) -> Result<()> {
+        let msg = SecureChannelMessage::Close;
+
+        // Encrypt the message
+        let msg = self.encrypt(ctx, msg).await?;
+
+        // Send the message to the decryptor on the other side
+        ctx.send_from_address(
+            self.remote_route.clone(),
+            msg,
+            self.addresses.encryptor.clone(),
+        )
+        .await?;
+
+        Ok(())
+    }
+
+    /// Schedule a DelayedEvent that will at specific point in time put a message
+    /// into EncryptorWorker's own internal mailbox which it will use as a trigger to get a new
+    /// credential and present it to the other side.
+    async fn schedule_credentials_refresh(&mut self, ctx: &Context, is_retry: bool) -> Result<()> {
+        let min_credential_expiration =
+            if let Some(min_credential_expiration) = self.min_credential_expiration {
+                min_credential_expiration
+            } else {
+                // Do nothing if there is no expiration
+                return Ok(());
+            };
+
+        // Cancel the old event
+        self.credential_refresh_event = None;
+
+        let now = now()?;
+
+        let duration = if min_credential_expiration < now + self.refresh_credential_time_gap {
+            Duration::from_secs(0)
+        } else {
+            // Refresh in self.refresh_credential_time_gap before the expiration
+            Duration::from_secs(
+                *(min_credential_expiration
+                    - self.refresh_credential_time_gap.as_secs().into()
+                    - now),
+            )
+        };
+
+        let duration = if duration < self.min_credential_refresh_interval && is_retry {
```

Comment:
That's a nitpick but I think that would read even better as 
```rust
// Avoid too many request to the credential_retriever, the refresh can't be sooner than
// self.min_credential_refresh_interval if it's a retry
let duration = 
    if is_retry {
        min(self.min_credential_refresh_interval, duration)
    } else { 
        duration 
    };

Reply:
You tricked me, should be `max` instead of `min` 😃

Reply:
yes, `max` :-)